src/scene/materials/shader-graph-node.js

```diff
@@ -0,0 +1,615 @@
+import { Texture } from '../../graphics/texture.js';
+import { Vec2 } from '../../math/vec2.js';
+import { Vec3 } from '../../math/vec3.js';
+import { Vec4 } from '../../math/vec4.js';
+
+var id = 0;
+
+/**
+ * @private
+ * @class
+ * @name ShaderGraphNode
+ * @classdesc A Shader Graph Node class used by shader graphs
+ * @param {string} funcGlsl - shader function used by this material
+ * @param {string} declGlsl - shader declarations used by the shader function
+ */
+var ShaderGraphNode = function (funcGlsl, declGlsl) {
+    this.name = "Untitled";
+    this.id = id++;
+
+    // storage for asset references
+    this._ioPortAssetReferences = [];
+    this._glslAssetReferences = {};
+    this._subGraphAssetReferences = [];
+
+    this.graphData = {};
+
+    this.graphData.ioPorts = []; // input, output or constant variables
+
+    if (funcGlsl) {
+        this.graphData.customFuncGlsl = funcGlsl;
+        this.graphData.customDeclGlsl = declGlsl;
+        this.genCustomFuncVars();
+    } else {
+        this.graphData.subGraphs = [];
+        this.graphData.connections = [];
+    }
+};
+
+ShaderGraphNode.prototype.constructor = ShaderGraphNode;
+
+Object.assign(ShaderGraphNode.prototype, {
+
+    clone: function () {
+        var clone = new ShaderGraphNode();
+
+        clone.graphData = {};
+
+        clone.graphData.ioPorts = this.graphData.ioPorts.slice(0);
+        clone.graphData.subGraphs = this.graphData.subGraphs.slice(0);
+        clone.graphData.connections = this.graphData.connections.slice(0);
+
+        clone.graphData.customDeclGlsl = this.graphData.customDeclGlsl;
+        clone.graphData.customFuncGlsl = this.graphData.customFuncGlsl;
+
+        return clone;
+    },
+
+    updateShaderGraphUniforms: function (mat) {
+        for (var n = 0; n < this.graphData.ioPorts.length; n++) {
+            var ioPort = this.graphData.ioPorts[n];
+
+            if (ioPort.name.startsWith('IN_') || (ioPort.name.startsWith('CONST_') && ioPort.type === 'sampler2D')) {

```

Comment:
ideally perhaps we have a list that allows us to avoid these string operations / testing here. We should have a list of uniforms to just loop over, and set parameter on material.

Reply:
yes - I will add _lists

Reply:
I've added optimizations remove/cache common string operations - however the main purpose of the class is to generate a glsl strings - so it doesn't make sense to go too far IMO